lib/event_summarizer/idv_matcher.rb

```diff
@@ -0,0 +1,436 @@
+
+module EventSummarizer
+  class IdvMatcher
+    IDV_WELCOME_SUBMITTED_EVENT = 'IdV: doc auth welcome submitted'
+    IDV_GPO_CODE_SUBMITTED_EVENT = 'IdV: enter verify by mail code submitted'
+    IDV_FINAL_RESOLUTION_EVENT = 'IdV: final resolution'
+    IDV_VERIFY_PROOFING_RESULTS_EVENT = 'IdV: doc auth verify proofing results'
+    IPP_ENROLLMENT_STATUS_UPDATED_EVENT = 'GetUspsProofingResultsJob: Enrollment status updated'
+    PROFILE_ENCRYPTION_INVALID_EVENT = 'Profile Encryption: Invalid'
+    RATE_LIMIT_REACHED_EVENT = 'Rate Limit Reached'
+
+    EVENT_PROPERTIES = ['@message', 'properties', 'event_properties'].freeze
+
+    VENDORS = {
+      'lexisnexis:instant_verify' => {
+        id: :instant_verify,
+        name: 'Instant Verify',
+        evaluator_module: EventSummarizer::VendorResultEvaluators::InstantVerify,
+      },
+      'aamva:state_id' => {
+        id: :aamva,
+        name: 'AAMVA',
+        evaluator_module: EventSummarizer::VendorResultEvaluators::Aamva,
+      },
+    }.freeze
+
+    UNKNOWN_VENDOR = {
+      id: :unknown,
+      name: 'Unknown vendor',
+    }.freeze
+
+    IdvAttempt = Data.define(
+      :started_at,
+      :significant_events,
+    ) do
+      def initialize(started_at:, significant_events: [])
+        super(started_at:, significant_events:)
+      end
+
+      def gpo?
+        self.significant_events.any? { |e| e.type == :start_gpo }
+      end
+
+      def ipp?
+        self.significant_events.any? { |e| e.type == :start_ipp }
+      end
+
+      def successful?
+        self.significant_events.any? do |e|
+          e.type == :identity_verified
+        end
+      end
+    end.freeze
+
+    SignificantIdvEvent = Data.define(
+      :timestamp,
+      :type,
+      :description,
+    ).freeze
+
+    attr_reader :current_idv_attempt
+    attr_reader :idv_attempts
+
+    # @return {Hash,nil}
+    def handle_cloudwatch_event(event)
+      @idv_attempts ||= []
+
+      case event['name']
+        when IDV_WELCOME_SUBMITTED_EVENT
+          start_new_idv_attempt(event:)
+
+        when IDV_FINAL_RESOLUTION_EVENT
+          for_current_idv_attempt(event:) do
+            handle_final_resolution_event(event:)
+          end
+
+        when IDV_GPO_CODE_SUBMITTED_EVENT
+          for_current_idv_attempt(event:) do
+            handle_gpo_code_submission(event:)
+          end
+
+        when IPP_ENROLLMENT_STATUS_UPDATED_EVENT
+          for_current_idv_attempt(event:) do
+            handle_ipp_enrollment_status_update(event:)
+          end
+
+        when IDV_VERIFY_PROOFING_RESULTS_EVENT
+          for_current_idv_attempt(event:) do
+            handle_verify_proofing_results_event(event:)
+          end
+
+        when PROFILE_ENCRYPTION_INVALID_EVENT
+          for_current_idv_attempt(event:) do
+            handle_profile_encryption_error(event:)
+          end
+
+        when RATE_LIMIT_REACHED_EVENT
+          handle_rate_limit_reached(event:)
```

Comment:
This does not use for_current_idv_attempt because not all rate limits are idv-related. We need to figure that out first.